tee/prototype/epoch.go

```diff
@@ -3,81 +3,198 @@
 package prototype
 
 import (
+	"errors"
+	"fmt"
 	"math/big"
 
 	"github.com/ethereum/go-ethereum/common"
-	log "github.com/sirupsen/logrus"
-
 	"github.com/perun-network/erdstall/tee"
+	log "github.com/sirupsen/logrus"
 )
 
-// An Epoch stores all balances of an epoch and contains signalling
-// infrastructure.
 type (
+	// Epoch manages deposits, transactions, and exits, as well as the
+	// progression of time in the system.
 	Epoch struct {
-		Number tee.Epoch
+		number tee.Epoch // Current deposit epoch.
+
+		exitLocked map[common.Address]struct{} // Withdrawing at end of epoch.
+		exitReqs   map[common.Address]struct{} // Requested exits.
+		balances   map[common.Address]*Acc     // Latest balances.
 
-		balances map[common.Address]*Bal
+		epochOutcome chan Outcome // The last epoch's outcome.
 	}
 
-	Bal struct {
+	// Acc contains an account's balance and transaction nonce.
+	Acc struct {
 		Nonce uint64
 		Value *big.Int
 	}
+
+	// Outcome contains all of an epoch's final balances, as well as all
+	// accounts that exited the system at the end of the epoch.
+	Outcome struct {
+		Epoch    tee.Epoch               // The finished epoch.
+		Exits    map[common.Address]*Acc // Exited accounts.
+		Accounts map[common.Address]*Acc // Final balances and nonces.
+	}
 )
 
+// NewEpoch creates an epoch manager that starts at the desired deposit epoch.
 func NewEpoch(n tee.Epoch) *Epoch {
 	return &Epoch{
-		Number:   n,
-		balances: make(map[common.Address]*Bal),
+		number:       n,
+		exitLocked:   make(map[common.Address]struct{}),
+		exitReqs:     make(map[common.Address]struct{}),
+		balances:     make(map[common.Address]*Acc),
+		epochOutcome: make(chan Outcome, 1),
 	}
 }
 
-// Next returns a clone of the current Epoch, with the Epoch counter increased
-// by one.
-func (e *Epoch) NewNext() *Epoch {
-	next := NewEpoch(e.Number + 1)
-	return next
+// DepositNum is the current deposit epoch number.
+func (e *Epoch) DepositNum() tee.Epoch { return e.number }
+
+// TxNum is the current transaction epoch number.
+func (e *Epoch) TxNum() tee.Epoch { return e.number - 1 }
+
+// ExitNum is the current exit epoch number.
+func (e *Epoch) ExitNum() tee.Epoch { return e.number - 2 }
+
+// Outcome has to be called exactly once after each phase shift and contains all
+// accounts that exited within this epoch, as well as all balances at the end of
+// the epoch. It blocks until the epoch in which it was last called ends.
+func (e *Epoch) Outcome() Outcome { return <-e.epochOutcome }
```

Comment:
Is there any gain in exposing it as
```suggestion
func (e *Epoch) Outcome() <-chan Outcome { return e.epochOutcome }
```
so that, at the calling site, we can `select` on some other event like enclave closing?

Reply:
no, it is only called internally in the block processor after the epoch has already shifted, but I guess it would make sense to add an assertion that it really exists, and panic instead of deadlocking in case it does not exist. The enclave cannot close at that point because it only after the block processor is done with the epoch in which a shutdown was requested.

Reply:
I added a log.Panic in case it is called more than once per epoch or before the epoch shift, because that would indicate a serious logic error.